### PR TITLE
rowcodec: make rowcodec can be used for mocktikv & unistore

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

first part of split PRs of #12634 (there are too many lines but almost them are test)

in #7597, we add rowcodec lib, but after see real-world usage old code need some change, and in this PR add much Unit Test to cover normal code path.

rowcodec provide:

1) encode:

- `Encode`: datumn array into row bytes, pay attention this method will flatten type and make int/unit as non-varint encode.

2) decoder:

- `DecodeToDatumMap`: decode row into datum map, mainly used in ddl operation in TiDB
- `DecodeToChunk`: decode row into chunk, mainly used in unistore executor and batch-get/point executor
- `DecodeToBytes`: decode row into old datum bytes, mainly used in TiDB memtable reader and mocktikv executor

`DecodeToBytes` can force varin via `WithVarint` option to decode int value as varint(in col value, pk value and default value), this will be used in mocktikv analyze columns executor to build table reader .

### What is changed and how it works?

- add new method and let method in they should be 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test test in https://github.com/pingcap/tidb/pull/12634 with unistore

Code changes

 - change lib and will be used in next pr

Side effects

 - n/a

Related changes

 - n/a

Release note

 - n/a

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/13774)

### CODE REVIEW

File: util/rowcodec/decoder.go [hunk 1]
```diff
@@ -29,133 +28,249 @@ import (
...
 	row
-	requestColIDs []int64
-	handleColID   int64
-	requestTypes  []*types.FieldType
-	origDefaults  [][]byte
-	loc           *time.Location
+	columns     []ColInfo
+	handleColID int64
+	loc         *time.Location
+	forceVarint bool
```
Comment (coocood) [inline util/rowcodec/decoder.go]: We can always encode to varint for DecodeToBytes and remove this option.

File: util/rowcodec/rowcodec_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2019 PingCAP, Inc.
+// Copyright 2019 PingCAP, Inc.	package rowcodec_test
```
Comment (coocood) [inline util/rowcodec/rowcodec_test.go]: accidental change

File: util/rowcodec/row.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	offsets []uint16
+	data    []byte
+
+	// for large row
+	colIDs32  []uint32
+	offsets32 []uint32
+}
+
+// String implements the strings.Stringer interface.
+func (r row) String() string {
```
Comment (coocood) [inline util/rowcodec/row.go]: This function is not needed.

File: util/rowcodec/decoder.go [hunk 4]
```diff
@@ -183,31 +298,163 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
+		}
+
+		idx, isNil, notFound := r.findColID(colID)
+		if !notFound && !isNil {
+			val := r.getData(idx)
+			values[offset] = decoder.encodeOldDatum(tp, val)
+			continue
+		}
+
+		if isNil {
```
Comment (jackysp) [inline util/rowcodec/decoder.go]: I think we should check the DefaultValue and HasNotNullFlag before it.

File: util/rowcodec/decoder.go [hunk 5]
```diff
@@ -183,31 +298,163 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
+			defVal, err := col.DefaultValue()
+			if err != nil {
+				return nil, err
+			}
+			if len(defVal) > 0 {
+				values[offset] = decoder.encodeOldDatum(tp, defVal)
+				continue
+			}
+		}
+		if mysql.HasNotNullFlag(uint(col.Flag)) {
```
Comment (jackysp) [inline util/rowcodec/decoder.go]: If the column not found and it doesn't have a default value, should we return an error directly? Why should we check the NotNullFlag?

File: util/rowcodec/decoder.go [hunk 6]
```diff
@@ -183,31 +298,163 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
+	case mysql.TypeYear:
+		flag = IntFlag
+	case mysql.TypeEnum, mysql.TypeBit, mysql.TypeSet:
+		flag = UintFlag
+	case mysql.TypeJSON:
+		flag = JSONFlag
+	case mysql.TypeNull:
+		flag = NilFlag
+	default:
+		err = errors.Errorf("unknown field type %d", tp)
```
Comment (jackysp) [inline util/rowcodec/decoder.go]: prefer to panic here

File: util/rowcodec/encoder.go [hunk 7]
```diff
@@ -113,203 +118,143 @@ func (encoder *Encoder) build(buf []byte) ([]byte, error) {
...
 
-func (encoder *Encoder) initColIDs() {
-	numCols := int(encoder.numNotNullCols + encoder.numNullCols)
-	if cap(encoder.colIDs) >= numCols {
-		encoder.colIDs = encoder.colIDs[:numCols]
-	} else {
-		encoder.colIDs = make([]byte, numCols)
+// EncodeFromOldRow encodes a row from an old-format row.
+// this method will be used in test.
+func encodeFromOldRow(encoder *Encoder, sc *stmtctx.StatementContext, oldRow, buf []byte) ([]byte, error) {
```
Comment (coocood) [inline util/rowcodec/encoder.go]: Can we define this function in test file?

File: util/rowcodec/decoder.go [hunk 8]
```diff
@@ -29,133 +29,242 @@ import (
...
-		handleColID:   handleColID,
-		requestTypes:  tps,
-		origDefaults:  xOrigDefaultVals,
-		loc:           sc.TimeZone,
-	}, nil
+	return row, nil
 }
 
-func convertDefaultValue(defaultVal []byte, sc *stmtctx.StatementContext) (colVal []byte, err error) {
+func (decoder *Decoder) decodeColDatum(col *ColInfo, colData []byte) (*types.Datum, error) {
```
Comment (coocood) [inline util/rowcodec/decoder.go]: It's more convenient to use value directly

File: util/rowcodec/decoder.go [hunk 9]
```diff
@@ -183,31 +292,139 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
+		buf = codec.EncodeUvarint(buf, decodeUint(val))
+	default:
+		buf = append(buf, tp)
+		buf = append(buf, val...)
+	}
+	return buf
+}
+
+// fieldType2Flag transforms field type into kv type flag.
+func fieldType2Flag(tp byte, f uint) (flag byte) {
```
Comment (coocood) [inline util/rowcodec/decoder.go]: `f` and `flag` are easy to be confused.

**Comment by lysu** [comment]
@jackysp @coocood I have make some refactor:

mainly split Decoder into DatumMapDecoder/ChunkDecoder/BytesDecoder, after refactor:

- decode chunk we can provider datum as default directly(avoid datum -> bytes ->  datum  cost)
- decode map we can avoid prepare default value, because decode map fill default value in logic in higher layer code
- decode bytes, we can provider bytes as default directly(avoid bytes -> datum -> bytes cost)

 PTAL if free

File: util/rowcodec/decoder.go [hunk 10]
```diff
@@ -14,148 +14,260 @@
...
-	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/codec"
 )
 
-// Decoder decodes the row to chunk.Chunk.
+// Decoder contains base util for decode row.
 type Decoder struct {
```
Comment (coocood) [inline util/rowcodec/decoder.go]: can we make it unexported?

File: util/rowcodec/decoder.go [hunk 11]
```diff
@@ -14,148 +14,260 @@
...
+
+		chk.AppendDatum(colIdx, &defDatum)
 	}
 	return nil
 }
 
-func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chunk) error {
-	ft := decoder.requestTypes[colIdx]
-	switch ft.Tp {
+func (decoder *ChunkDecoder) decodeColChunk(colIdx int, col *ColInfo, colData []byte, chk *chunk.Chunk) error {
```
Comment (coocood) [inline util/rowcodec/decoder.go]: `decodeColToChunk`

File: util/rowcodec/decoder.go [hunk 12]
```diff
@@ -183,31 +295,161 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
 }
+
+// BytesDecoder decodes the row to old datums bytes.
+type BytesDecoder struct {
+	Decoder
+	defBytes []func() ([]byte, error)
+}
+
+// NewByteDecoder creates a BytesDecoder.
+func NewByteDecoder(columns []ColInfo, handleColID int64, defBytes []func() ([]byte, error), loc *time.Location) *BytesDecoder {
```
Comment (coocood) [inline util/rowcodec/decoder.go]: We need  comments about the `defBytes` argument.

File: util/rowcodec/decoder.go [hunk 13]
```diff
@@ -183,31 +295,158 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
...
-		return errors.Errorf("unknown type %d", ft.Tp)
+		return errors.Errorf("unknown type %d", col.Tp)
 	}
 	return nil
 }
+
+// BytesDecoder decodes the row to old datums bytes.
+type BytesDecoder struct {
+	decoder
+	defBytes []func() ([]byte, error)
```
Comment (coocood) [inline util/rowcodec/decoder.go]: If we make the `defBytes` accept an `i` argument, then we only need to implement one function.

**Comment by coocood** [comment]
LGTM

**Comment by lysu** [comment]
PTAL @jackysp  :smile:

**Comment by lysu** [comment]
PTAL @jackysp

**Pull request merged**